dotnet/src/Agents/OpenAI/OpenAIAssistantChannel.cs

```diff
@@ -324,48 +394,86 @@ private static ChatMessageContent GenerateImageFileContent(string agentName, Aut
         return messageContent;
     }
 
-    private static IEnumerable<Task<ToolOutput>> ExecuteStep(OpenAIAssistantAgent agent, RunStep step, CancellationToken cancellationToken)
+    private static ChatMessageContent GenerateCodeInterpreterContent(string agentName, RunStepCodeInterpreterToolCall contentCodeInterpreter)
     {
-        // Process all of the steps that require action
-        if (step.Status == RunStepStatus.InProgress && step.StepDetails is RunStepToolCallDetails callDetails)
-        {
-            foreach (RunStepFunctionToolCall toolCall in callDetails.ToolCalls.OfType<RunStepFunctionToolCall>())
+        return
+            new ChatMessageContent(
+                AuthorRole.Tool,
+                [
+                    new TextContent(contentCodeInterpreter.Input)
+                ])
             {
-                // Run function
-                yield return ProcessFunctionStepAsync(toolCall.Id, toolCall);
-            }
+                AuthorName = agentName,
+            };
+    }
+
+    private static ChatMessageContent GenerateFunctionCallContent(string agentName, FunctionCallContent[] functionSteps)
+    {
+        ChatMessageContent functionCallContent = new(AuthorRole.Tool, content: null)
+        {
+            AuthorName = agentName
+        };
+
+        functionCallContent.Items.AddRange(functionSteps);
+
+        return functionCallContent;
+    }
+
+    private static ChatMessageContent GenerateFunctionResultContent(string agentName, FunctionCallContent functionStep, string result)
+    {
+        ChatMessageContent functionCallContent = new(AuthorRole.Tool, content: null)
+        {
+            AuthorName = agentName
+        };
+
+        functionCallContent.Items.Add(
+            new FunctionResultContent(
+                functionStep.FunctionName,
+                functionStep.PluginName,
+                functionStep.Id,
+                result));
+
+        return functionCallContent;
+    }
+
+    private static Task<FunctionResultContent>[] ExecuteFunctionSteps(OpenAIAssistantAgent agent, FunctionCallContent[] functionSteps, CancellationToken cancellationToken)
+    {
+        Task<FunctionResultContent>[] functionTasks = new Task<FunctionResultContent>[functionSteps.Length];
+
+        for (int index = 0; index < functionSteps.Length; ++index)
+        {
+            functionTasks[index] = ExecuteFunctionStepAsync(functionSteps[index]);
         }
 
-        // Local function for processing the run-step (participates in method closure).
-        async Task<ToolOutput> ProcessFunctionStepAsync(string callId, RunStepFunctionToolCall functionDetails)
+        return functionTasks;
+
+        async Task<FunctionResultContent> ExecuteFunctionStepAsync(FunctionCallContent functionStep)
```

Comment:
nit: it seems the method turned into one liner and can be inlined.

Reply:
Updated